h/services/url_migration.py

```diff
@@ -0,0 +1,117 @@
+import copy
+import logging
+
+import h.storage
+from h.events import AnnotationEvent
+from h.exceptions import RealtimeMessageQueueError
+from h.models import Annotation, DocumentURI
+from h.schemas.annotation import _document as transform_document
+from h.util.uri import normalize
+
+log = logging.getLogger(__name__)
+
+
+class URLMigrationService:
+    """
+    Moves annotations from one URL to another.
+    """
+
+    def __init__(self, request):
+        self.request = request
+
+    def update_annotation_urls(self, url_mappings):
+        """
+        Migrate annotations from one set of URLs to another.
+
+        :param url_mappings: Dict of URL mappings that conform to `URLMigrationSchema`.
+        """
+
+        session = self.request.db
+
+        for url, updates in url_mappings.items():
+            uri_normalized = normalize(url)
+
+            annotations = (
+                session.query(Annotation)
+                .join(DocumentURI, Annotation.document_id == DocumentURI.document_id)
+                .filter(DocumentURI.uri_normalized == uri_normalized)
+            )
+
+            processed_count = 0
+
+            for ann in annotations:
+                ann_update_data = {
+                    "target_uri": updates["url"],
+                }
+
+                if "document" in updates:
+                    ann_update_data["document"] = transform_document(
+                        updates["document"], updates["url"]
+                    )
+
+                # Add selectors to annotation if there is no selector of the same
+                # type.
+                #
+                # This change is specifically to aid in the migration of ebook
+                # annotations from a chapter/page URL to the containing book.
+                # The information about which chapter/page the annotation refers
+                # to is then moved into selectors.
+                #
+                # See https://github.com/hypothesis/h/issues/7709
+                if new_selectors := updates.get("selectors"):
+                    selectors = copy.deepcopy(ann.target_selectors)
+                    for new_sel in new_selectors:
+                        if not any(s for s in selectors if s["type"] == new_sel.type):
+                            selectors.append(new_sel)
+                    ann.target_selectors = selectors
+
+                # Update the annotation's `target_uri` and associated document,
+                # and create `Document*` entities for the new URL if they don't
+                # already exist.
+                h.storage.update_annotation(
+                    self.request,
+                    ann.id,
+                    ann_update_data,
+                    # Don't update "edited" timestamp on annotation cards.
+                    update_timestamp=False,
+                )
```

Comment:
Oh, I think you also need to pass `force=True` to the `add_by_id()` call that `storage.update_annotation()` makes, which means you'll need to add an optional `force` argument to `update_annotation()` as well. I don't think the code is working currently:

* `storage.update_annotation()` will [add annotations to the job queue](https://github.com/hypothesis/h/blob/26250b0fbc09430c1aef4d5972ae711ae9987829/h/storage.py#L192-L194) with `force=False` (see the default value of `force=False` in [`Queue.add_by_id()`](https://github.com/hypothesis/h/blob/26250b0fbc09430c1aef4d5972ae711ae9987829/h/services/search_index/_queue.py#L92-L102))
* What `force=True` does is it [forces re-indexing of an annotation even if its Elasticsearch already appears to be up to date](https://github.com/hypothesis/h/blob/26250b0fbc09430c1aef4d5972ae711ae9987829/h/services/search_index/_queue.py#L177-L189)
* With `force=False` the periodic reindexing task will [only reindex the annotation if the current document in Elasticsearch is not "equal" to the annotation in Postgres](https://github.com/hypothesis/h/blob/26250b0fbc09430c1aef4d5972ae711ae9987829/h/services/search_index/_queue.py#L204-L219), otherwise it will simply mark the job in the DB as completed without doing anything. "Equal" is defined as [if the Elasticsearch document has the same `user` and `updated` as the Postgres row](https://github.com/hypothesis/h/blob/26250b0fbc09430c1aef4d5972ae711ae9987829/h/services/search_index/_queue.py#L280-L286). Since you're not updating `updated` your annotations will be seen as equal and will not be reindexed

I think the reason why your annotations **are** being reindexed in your testing is because your service is also calling `search_index.handle_annotation_event()` (synchronous reindexing) below. So you're reindexing the annotation twice: once synchronously and once asynchronously. And the asynchronous attempt isn't working.

You definitely want to use the job queue/asynchronous reindexing for this, not the synchronous reindexing. See my comments on the `search_index.handle_annotation_event()` call below. The job queue/asynchronous reindexing **guarantees** that your annotation will be updated in Elasticsearch (it will keep on trying until it successfully reads back the correctly updated data from Elasticsearch) whereas the synchronous reindexing is "fire and forget" and does not guarantee that the reindex actually worked. You definitely want the "slower but guaranteed" version for batch reindexes like this

Reply:
Thanks for the info. I started making changes along these lines but got confused / tripped up by several things along the way:

1. At first, the reindexing tasks did not get processed, because I didn't have `h_periodic` running locally.
2. When `h_periodic` is running normally, it only runs `sync_annotations` once every minute, which is fine for the automatic reindexing but is quite confusing if you're trying to use the admin console to trigger reindexing manually.
3. As I hadn't run h_periodic in a long time (maybe not ever) locally, my local h DB had a large volume of expired entries in the job queue, and it looks like [nothing removes them](https://hypothes-is.slack.com/archives/C1MA4E9B9/p1673873630354619)